File: graph/disjoint_set.go
package graph

type DisjointSet struct {
	parents []int
}

func NewDisjointSet(n int) *DisjointSet {
	res := DisjointSet{
		parents: make([]int, n),
	}

	for i := 0; i < n; i++ {
		res.parents[i] = -1 // initialize the union size to 1 (negative value to store the size of the union, while also saying that it's a union leaderid)
	}

	return &res
}

func (d *DisjointSet) Find(x int) int {
	if d.parents[x] < 0 {
		return x
	}

	path := make([]int, 0)

	for d.parents[x] >= 0 {
		path = append(path, x)
		x = d.parents[x]
	}

	for _, p := range path {
		d.parents[p] = x
	}

	return x
}

func (d *DisjointSet) Union(a, b int) bool {
	if a == b {
		return false
	}

	aId, bId := d.Find(a), d.Find(b)

	if aId == bId {
		return false
	}

	aSize, bSize := d.GetUnionSize(aId), d.GetUnionSize(bId)

	if bSize > aSize {
		d.parents[aId] = bId
		d.parents[bId] = -(aSize + bSize)
	} else {
		d.parents[bId] = aId
		d.parents[aId] = -(aSize + bSize)
	}
	return true
}

func (d *DisjointSet) GetUnionSize(a int) int {
	return -d.parents[d.Find(a)]
}

func (d *DisjointSet) GetTotalUnions() int {
	var count int
	for i := 0; i < len(d.parents); i++ {
		if d.parents[i] < 0 {
			count += 1
		}
	}
	return count
}

File: src/disjoint_set.rs
use std::vec::Vec;

pub struct DisjointSet {
  parents: Vec<i32>,
}

impl DisjointSet {
  fn new(n: u32) -> Self {
    return DisjointSet { 
      parents: vec![-1; n as usize], 
    }
  }

  fn find(&mut self, x: u32) -> u32 {
    let mut x = x as usize;
    if self.parents[x] < 0 {
      return x as _;
    }
    

    let mut path = vec![];
    while self.parents[x] >= 0 {
      path.push(x);
      x = self.parents[x] as usize;
    }

    for &p in path.iter() {
      self.parents[p] = x as i32;
    }
    return x as _;
  }

  fn union(&mut self, a: u32, b: u32) -> bool {
    if a == b {
      return false;
    }

    let (a_id, b_id) = (self.find(a) as i32, self.find(b) as i32);

    if a_id == b_id {
      return false
    }

    let (a_size, b_size) = (self.union_size(a_id as u32), self.union_size(b_id as u32));

    if b_size > a_size {
      self.parents[a_id as usize] = b_id as i32;
      self.parents[b_id as usize] = -((a_size+b_size) as i32);
    }else{
      self.parents[b_id as usize] = a_id as i32;
      self.parents[a_id as usize] = -((a_size+b_size) as i32);
    }
    return true
  }

  fn union_size(&mut self, x: u32) -> u32 {
    let x_id = self.find(x);
    return (-self.parents[x_id as usize]) as u32
  }

  fn total_set(&self) -> u32 {
    let mut result = 0;
    for &i in self.parents.iter() {
      if i < 0 {
        result += 1;
      }
    }
    return result as u32;
  }
}

#[test]
fn test_disjoint_set() {
  let mut ds = DisjointSet::new(3);
  assert_eq!(1, ds.find(1));
  assert_eq!(3, ds.total_set());

  ds.union(1, 2);
  assert_eq!(ds.find(1), ds.find(2));
  assert_eq!(2, ds.union_size(1));
  assert_ne!(ds.find(0), ds.find(1));
  assert_eq!(2, ds.total_set());
  
}